### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive for `throw` branch

### ISSUE THREAD

**Issue by JounQin** [typescript-eslint/typescript-eslint#11089]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

[Playground](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgOUzAQEkl0I4BvAKDjgCNlMoBPQgW2AC44BnGFGQBzOnDA4A1pmHBOPfoJFiYLWTD68BQpMIDaAXTEB3THw4B%2BXrXr0GUCMb5or1Mbbjo%2BrhhAgAbYEwkdzgAXzEIiJpQSFhEFDQsXDgANUJiMgo4UFQkABM%2BAiJSckobOAA3ND4ECCQtJV0aaNjoeGRUDBw8AAUpGWAAKT56t3okTC5GnVF6CDAYOsn-ABFgMGACrewEYG9eACVgbGh8gB5tEQAaRVmAPjFJ4lcM0ooWmhjwds8AVyQ2CWY3UbyyEAAYg4OP1sNJZCN6gAKMQSOGDRENOCw%2BHDUZIa40ACUvHSJXB4zgpyQAmocGeCFuC2BK3Wm22gL2RTCcAAvOIBgj8WIEOg4EiAIQMuAAHxlcAlzOWmDWGy2%2BR2XKJlPoMAAFg5jPTgEaAKJQBxQFEeegAcnwlAABgzHfToHBHUr6iq2erNftXeg9v58p4IADQ8gPWjcQA6ABW%2BMdsexgTMeC2fD%2BUDw%2BrwMcGiCKDD%2BCH88GMCH1BAAgr0SPw-mA4jBY7bCR4iZExIF4NUoLV6jMRLK4BHgEGkMB8iKxdbbFKSrHVFB1HxQvRYxwiEiV%2Bo%2BfcPQASKgM2MF2TyMIAWhPe%2BAMDCjq7NrgseA-dYSLAkmEB51NrUrSF7AKkNTLHycBeqyaocrs%2Bx6D%2BwhGK%2B9CiuK-aDkg2oVKhiBzphEESry-IgWBA7LDhG6ofqhrGmaFrQPOeEeI6ZBAQgAhbPAIFVOB9RFBQEZhlAcAniBmKxpMXA8pW1bib%2BYSxo6HYsfQL4sREeE8sA-jOABr6EWMpGCqB-EhNp1H0DmMDZkgcCCH8wDUWE2pyqEEpGRZcBUR4tGOPRcDmpazG2GxgICZxeQ8aZfEUQJYbCRQoniaZknScASkqaEGlaXANl2QBXmqZusZSSUqlRDQQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

### Repro Code

```TypeScript
export interface NapiInfo {
  binaryName: string
  packageName: string
  targets: string[]
  wasm?: {
    browser?: {
      fs?: boolean
    }
  }
}

export interface VNapiInfo extends NapiInfo {
  version: string
}

export interface PackageJson {
  name: string
  optionalDependencies?: Record<string, string>
  napi?: NapiInfo
}

export function getNapiInfoFromPackageJson(
  packageJson: PackageJson,
): VNapiInfo {
  const { napi, optionalDependencies } = packageJson
  if (!napi || !optionalDependencies) {
    throw new Error(
      'No `napi` nor `optionalDependencies` field found in `package.json`. Please ensure the package is built with NAPI support.',
    )
  }
  let version: string | undefined
  if (
    !napi.targets
      .map(target => `${napi.packageName}-${target}`)
      .every(pkg => {
        const packageVersion = optionalDependencies[pkg]
        if (version) {
          if (version !== packageVersion) {
            throw new Error(
              `Inconsistent package versions found for ${packageJson.name} with ${pkg}.`,
            )
          }
        } else {
          version = packageVersion
        }
        return true
      }) ||
    !version
  ) {
    throw new Error(
      `Inconsistent package versions found for ${packageJson.name}.`,
    )
  }
  return {
    version,
    ...napi,
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error

### Actual Result

Error reported

**Comment by JounQin** [comment]
This is just for demo, I changed to use `forof` instead.

**Comment by bradzacher** [comment]
I'm not quite sure I understand - how is this a false positive? The function always returns `true`. This means it's never filtering anything. To me it looks like a completely valid flagging.

You could just as easily refactor the code to not use `every` by hoisting it outside of the condition to a stand-alone loop and the overall logic and control flow of the function would be the same.

**Comment by JounQin** [comment]
> You could just as easily refactor the code to not use every by hoisting it outside of the condition to a stand-alone loop and the overall logic and control flow of the function would be the same.

Yes, that's what I'm doing now.

---

I was thinking `throw` should be considered as a different branch with the consistent `return true` as it has different affects with return statement, but indeed it can simplified as described above, so if you it's a won't fix, that's understandable.

**Comment by kirkwaiblinger** [comment]
After squinting, I understand why this code is useful, but I agree with "working as intended"/"wontfix"... Let's say we simplify the repro to something like the following:
([playground](https://typescript-eslint.io/play/#ts=5.8.2&fileType=.tsx&code=GYVwdgxgLglg9mABMAFAQwE4YFyLCAWwCMBTDAbQF0BKRAbwChFEIEBnKRNuAkgYQQATGLASIAvIkwYAdCQBuZAJ4oAHhIB89Js0QxgiNYgA8iAAy1Gu3VAAWGOAHc8JZwFEscDCmoBuHcwAvgGIIRgkUCAYSFAYICT%2BQX4MOvqG3LwCYMKiYJYhrGDcADYkMsVwAOYoAOSx8YhEGGiQtjXJQYgkxWwk2taFJWUV1TXAaD19TS0QbR2IwYFAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRQA7iKaRwYAL4hNQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false))
```ts
function f(arr: number[]) {
  const someCondition = arr.every(x => {
    if (x < 0) {
      throw new Error();
    }
    
    return true;
  });

  if (someCondition) {
    console.log('true branch');
  } else {
    console.log('false branch');
  }
}
```

In this example it's clearly a misuse of `.every` and the linter report in question is actually the only clue that the false branch is unreachable.

Clearly either 
```ts
function f(arr: number[]) {
  const someCondition = arr.every(x => {
    if (x < 0) {
      return false;
    }
    
    return true;
  });

  if (someCondition) {
    console.log('true branch');
  } else {
    console.log('false branch');
  }
}
```

or 

```ts
function f(arr: number[]) {
  arr.forEach(x => {
    if (x < 0) {
      throw new Error();
    }
  });

  console.log('true branch');
}
```

is preferable. So the report here is pretty hard to argue with.

However, the one thing that the  always-truthy `.every()` allows you to do is avoid an intermediate variable in some situations, like the following.
```ts
function f(arr: number[]) {
  return arr
    .filter(x => someFilter(x))
    // can't be converted to forEach, since forEach returns undefined instead of the array
    .every(x => {
      importantSideEffect(x);
      return true;
    })
    .map(x => someMap(x));
}
```

But I think that an intermediate variable is a reasonable price to pay for the added clarity in the control flow:
```ts
function f(arr: number[]) {
  const filtered = arr.filter(x => someFilter(x));
  for (const x of filtered) {
      importantSideEffect(x);   
  }
  return filtered.map(x => someMap(x));
}
```

Therefore, IMO it's ok to say that this rule explicitly does not support using `.every()` as a control flow mechanism and this report is "working as intended". But ACK that that's a slightly opinionated judgement call.

**Comment by kirkwaiblinger** [comment]
FWIW, I tested out the refactoring on the OP, and this seems this seems much more obvious to read, no?
```ts
export interface NapiInfo {
  binaryName: string
  packageName: string
  targets: string[]
  wasm?: {
    browser?: {
      fs?: boolean
    }
  }
}

export interface VNapiInfo extends NapiInfo {
  version: string
}

export interface PackageJson {
  name: string
  optionalDependencies?: Record<string, string>
  napi?: NapiInfo
}

export function getNapiInfoFromPackageJson(
  packageJson: PackageJson,
): VNapiInfo {
  const { napi, optionalDependencies } = packageJson
  if (!napi || !optionalDependencies) {
    throw new Error(
      'No `napi` nor `optionalDependencies` field found in `package.json`. Please ensure the package is built with NAPI support.',
    )
  }

  const pkgs = napi.targets.map(target => `${napi.packageName}-${target}`);

  let version: string | undefined
  for (const pkg of pkgs) {
    const packageVersion = optionalDependencies[pkg]
    if (version) {
      if (version !== packageVersion) {
        throw new Error(
          `Inconsistent package versions found for ${packageJson.name} with ${pkg}.`,
        )
      }
    } else {
      version = packageVersion
    }
  }
  
  if (
    !version
  ) {
    throw new Error(
      `Inconsistent package versions found for ${packageJson.name}.`,
    )
  }
  return {
    version,
    ...napi,
  }
}
```

**Comment by JounQin** [comment]
@kirkwaiblinger Yes, that's what I'm doing now:

https://github.com/un-ts/napi-postinstall/blob/bbdef3a10c69c856b47175842eae39374cce7a13/src/helpers.ts#L43-L105
